relayer/src/supervisor/scan.rs

```diff
@@ -0,0 +1,774 @@
+            }
+        };
+
+        Ok(scan)
+    }
+
+    pub fn query_allowed_channels(
+        &mut self,
+        chain: &Chain,
+        spec: &ChannelsSpec,
+        scan: &mut ChainScan,
+    ) -> Result<(), Error> {
+        info!("querying allowed channels...");
+
+        for (port_id, channel_id) in spec.iter() {
+            let result = scan_allowed_channel(self.registry, chain, port_id, channel_id);
+
+            match result {
+                Ok(ScannedChannel {
+                    channel,
+                    counterparty_channel,
+                    connection,
+                    counterparty_connection_state,
+                    client,
+                }) => {
+                    let client_scan = scan
+                        .clients
+                        .entry(client.client_id.clone())
+                        .or_insert_with(|| ClientScan::new(client));
+
+                    let connection_scan = client_scan
+                        .connections
+                        .entry(connection.connection_id.clone())
+                        .or_insert_with(|| {
+                            ConnectionScan::new(connection, counterparty_connection_state)
+                        });
+
+                    connection_scan
+                        .channels
+                        .entry(channel.channel_id.clone())
+                        .or_insert_with(|| ChannelScan::new(channel, counterparty_channel));
+                }
+                Err(e) => error!("failed to scan channel '{}', reason: {}", channel_id, e),
+            }
+        }
+
+        Ok(())
+    }
+
+    pub fn scan_all_clients(&mut self, chain: &Chain, scan: &mut ChainScan) -> Result<(), Error> {
+        info!("scanning all clients...");
+
+        let clients = query_all_clients(chain)?;
+
+        for client in clients {
+            if let Some(client_scan) = self.scan_client(chain, client)? {
+                scan.clients.insert(client_scan.id().clone(), client_scan);
+            }
+        }
+
+        Ok(())
+    }
+
+    fn scan_client(
+        &mut self,
+        chain: &Chain,
+        client: IdentifiedAnyClientState,
+    ) -> Result<Option<ClientScan>, Error> {
+        let span = info_span!("scan.client", client = %client.client_id);
+        let _guard = span.enter();
+
+        info!("scanning client...");
+
+        if !self.client_allowed(chain, &client) {
+            warn!(
+                trust_threshold = ?client.client_state.trust_threshold(),
+                "skipping client, reason: client is not allowed",
+            );
+
+            return Ok(None);
+        }
+
+        let counterparty_chain_id = client.client_state.chain_id();
+        let has_counterparty = self.config.has_chain(&counterparty_chain_id);
+
+        if !has_counterparty {
+            debug!(
+                chain_id = %chain.id(),
+                counterparty_chain_id = %counterparty_chain_id,
```

Comment:
There may be a need for more consistency in key names for the chains. In some other places in #1491, we used `src_chain` and `dst_chain`.

Reply:
At this stage neither chain is a source of a destination, they're just counterparts to each other so I would say that it makes more sense to name them as such. In other places, such as when relaying packets, then the chains can be seen as a source and a destination. Wonder what @adizere thinks?

Reply:
I see. I think this is good then, but the key names could be shortened to `chain` and `counterparty_chain`.

Reply:
496d735